### PR TITLE
Updated hamming to check if size of w does not match the size of u and v

### LINKED ISSUE #17725: BUG: spatial: Bad error message from `hamming` when `w` has the wrong size.
The error message reported by `scipy.spatial.distance.hamming` when the size of the weights parameter `w` does not match the size of the input vectors is not correct.  The following is with the main developement branch and numpy 1.24.0:

```
In [1]: import scipy

In [2]: scipy.__version__
Out[2]: '1.11.0.dev0+1231.71c19cc'

In [3]: import numpy as np

In [4]: np.__version__
Out[4]: '1.24.0'

In [5]: from scipy.spatial.distance import hamming
```

Call `hamming` with a weight array whose size does not match the sizes of `u` and `v`:

```
In [6]: hamming([1, 0, 1, 1], [1, 1, 0, 1], w=[1, 3, 1])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [6], line 1
----> 1 hamming([1, 0, 1, 1], [1, 1, 0, 1], w=[1, 3, 1])

File ~/repos/git/forks/scipy/build-install/lib/python3.10/site-packages/scipy/spatial/distance.py:720, in hamming(u, v, w)
    718 if w is not None:
    719     w = _validate_weights(w)
--> 720 return np.average(u_ne_v, weights=w)

File <__array_function__ internals>:200, in average(*args, **kwargs)

File ~/py3.10.8/lib/python3.10/site-packages/numpy/lib/function_base.py:532, in average(a, axis, weights, returned, keepdims)
    530 if a.shape != wgt.shape:
    531     if axis is None:
--> 532         raise TypeError(
    533             "Axis must be specified when shapes of a and weights "
    534             "differ.")
    535     if wgt.ndim != 1:
    536         raise TypeError(
    537             "1D weights expected when shapes of a and weights differ.")

TypeError: Axis must be specified when shapes of a and weights differ.

```
The error message should just say that the size of `w` does not match the size of `u` and `v`.  The reference to an `axis` parameter is misleading, because that error is actually being raised by a function called by `hamming`; `hamming` does not have an `axis` parameter.  I suspect `hamming` should do some parameter validation earlier than it currently does.

### PR DESCRIPTION
**Pull request by aadya940**
#### Reference issue
Closes #17725 .

#### What does this implement/fix?
Added an if-condition to check if shape of weights does not match the shape of inputs, if true, an error is raised.

### CODE REVIEW

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -717,7 +717,10 @@ def hamming(u, v, w=None):
         raise ValueError('The 1d arrays must have equal lengths.')
     u_ne_v = u != v
     if w is not None:
-        w = _validate_weights(w)
+         if (w.shape == u.shape)
+            w = _validate_weights(w)
+         else:
+            raise ValueError('w should have same dimensions as u and v')
```
Comment (j-bowhay) [inline scipy/spatial/distance.py:723]: The indentation here is not correct

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Thanks @aadya940, this will also need a test that checks the error is raised

**Comment by aadya940** [comment]
> Thanks @aadya940, this will also need a test that checks the error is raised

Hey @j-bowhay ,
I have submitted the code with a test. Please do check it and help me out as I'm new to contributing to Scipy and unit-testing. Thanks.

**Comment by j-bowhay** [comment]
Please could you test this locally (https://docs.scipy.org/doc/scipy/dev/contributor/development_workflow.html) and fix the syntax errors then I will take a look :)

File: scipy/spatial/distance.py [hunk 2]
```diff
@@ -717,7 +717,10 @@ def hamming(u, v, w=None):
         raise ValueError('The 1d arrays must have equal lengths.')
     u_ne_v = u != v
     if w is not None:
-        w = _validate_weights(w)
+      if (w.shape == u.shape):
+         w = _validate_weights(w)
+       else:
```
Comment (melissawm) [inline scipy/spatial/distance.py:722]: I believe you have an indentation error here?

File: scipy/spatial/tests/test_distance.py [hunk 3]
```diff
@@ -1934,6 +1934,11 @@ def test_hamming_unequal_length():
     # Used to give an AttributeError from ndarray.mean called on bool
     assert_raises(ValueError, whamming, x, y)
 
+def test_hamming_weights_length_mismatch():
+    x = [1, 0, 1, 1]
+    y = [1, 1, 0, 1]
+    w = [1, 3, 1]
+    assert_raises(ValueError, x, y, w)
```
Comment (melissawm) [inline scipy/spatial/tests/test_distance.py:1941]: It doesn't look like this test is testing anything - maybe you forgot something?

**Comment by j-bowhay** [comment]
Thanks @aadya940, the original issue has now been closed so I am going to close this. Regardless, thank you for your contribution

**Pull request closed without merging**